src/drt/src/pa/FlexPA_prep.cpp

```diff
@@ -2040,48 +2036,41 @@ void FlexPA::genInstRowPattern_print(std::vector<FlexDPNode>& nodes,
   }
 }
 
-int FlexPA::getEdgeCost(const int prev_node_idx,
-                        const int curr_node_idx,
-                        const std::vector<FlexDPNode>& nodes,
+int FlexPA::getEdgeCost(FlexDPNode* prev_node,
+                        FlexDPNode* curr_node,
                         const std::vector<frInst*>& insts)
 {
-  int edge_cost = 0;
-  int prev_idx_1, prev_idx_2, curr_idx_1, curr_idx_2;
-  getNestedIdx(
-      prev_node_idx, prev_idx_1, prev_idx_2, ACCESS_PATTERN_END_ITERATION_NUM);
-  getNestedIdx(
-      curr_node_idx, curr_idx_1, curr_idx_2, ACCESS_PATTERN_END_ITERATION_NUM);
-  if (prev_idx_1 == -1 || curr_idx_1 == (int) insts.size()) {
-    return edge_cost;
+  auto [prev_inst_idx, prev_acc_pattern_idx] = prev_node->getIdx();
+  auto [curr_inst_idx, curr_acc_pattern_idx] = curr_node->getIdx();
+  if (prev_node->isSource() || curr_node->isDrain()) {
```

Comment:
use isVirtual for order independence. (no cost between a virtual node and any other node)